subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/compilerapi/constants/ConstantToDependentsMapping.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.incremental.compilerapi.constants;
+
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import it.unimi.dsi.fastutil.ints.IntSet;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import org.gradle.api.internal.cache.StringInterner;
+import org.gradle.internal.serialize.AbstractSerializer;
+import org.gradle.internal.serialize.Decoder;
+import org.gradle.internal.serialize.Encoder;
+import org.gradle.internal.serialize.IntSetSerializer;
+import org.gradle.internal.serialize.InterningStringSerializer;
+import org.gradle.internal.serialize.ListSerializer;
+import org.gradle.internal.serialize.MapSerializer;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A data structure that holds constant to class mapping.
+ * It consists of dependents class names `dependentClasses` and dependents `privateConstantDependents` and `accessibleConstantDependents` as `constant hash` => set(indexes) mapping.
+ *
+ * The idea here is that Strings take a lot of memory and there is one-to-many relationship between constant to dependents.
+ * So that is why we store String names of dependents in a separate list and we access them via indexes.
+ **/
+public class ConstantToDependentsMapping {
+
+    /**
+     * Visited classes will not be serialized and accessible after reading from cache.
+     * They are needed just as output of Compiler plugin to correctly merge new mapping to old one.
+     */
+    private transient final Set<String> visitedClasses;
+
+    private final List<String> dependentClasses;
+    private final Map<String, Set<Integer>> privateConstantDependents;
```

Comment:
:o: : I think this should always use IntSet, not any arbitrary Set. Both for clarity and to ensure efficiency.

Reply:
+1

Reply:
fixed